src/allmydata/immutable/layout.py

```diff
@@ -381,10 +389,15 @@ def _parse_offsets(self, data):
             self._offsets[field] = offset
         return self._offsets
 
-    def _fetch_sharehashtree_and_ueb(self, offsets):
+    @async_to_deferred
+    async def _fetch_sharehashtree_and_ueb(self, offsets):
+        [ueb_length] = struct.unpack(
+            await self._read(offsets['share_hashes'], self._fieldsize),
+            self._fieldstruct
```

Comment:
As far as I can tell `_fetch_sharehashtree_and_ueb` and `_parse_sharehashtree_and_ueb` are both dead code written in preparation the optimization of reading past the end of the share instead of reading exactly the right length.  Since we've decided to rule out allowing reading past the end of the share, this optimization is no longer possible.

So perhaps we just delete these two methods (and the commented out code that would have called them).

This also means that the two "..._the_old_way" methods are actually just "the way" so possibly we could rename them (but this doesn't seem urgent).  Also the `_share_hashes` and `_ueb_data` attributes can't ever be set (since they are only set by the dead code mentioned above) so `get_share_hashes` and `_get_uri_extension` are really just exact aliases for `_get_share_hashes_the_old_way` and `_get_uri_extension_the_old_way`, respectively...

Reply:
Ooh nice.